File: LeetCode/Rust/977.squares-of-a-sorted-array.rs
/*
 * @lc app=leetcode id=977 lang=rust
 *
 * [977] Squares of a Sorted Array
 */

// @lc code=start
impl Solution {
    pub fn sorted_squares(nums: Vec<i32>) -> Vec<i32> {
        let mut square_nums: Vec<i32> = Vec::new();
        // TODO: Improve
        for num in nums.iter() {
            square_nums.push(num * num)
        }
        square_nums.sort();


        square_nums
    }
}
// @lc code=end

File: LeetCode/CPP/977.squares-of-a-sorted-array.cpp
#include <vector>
#include <iostream>
using namespace std;
/*
 * @lc app=leetcode id=977 lang=cpp
 *
 * [977] Squares of a Sorted Array
 */

// @lc code=start
class Solution
{
public:
    vector<int> sortedSquares(vector<int> &nums)
    {
        vector<int> arr(nums.size());
        // Take two pointers
        int i = 0, j = nums.size() - 1;
        for (int c = arr.size() - 1; c >= 0; c--)
        {
            // Fill the new array depending on the bigger absolute value
            if (abs(nums[i]) > abs(nums[j]))
            {
                arr[c] = nums[i] * nums[i];
                i++;
            }
            else
            {
                arr[c] = nums[j] * nums[j];
                j--;
            }
        }
        return arr;
    }
};
// @lc code=end
